### DISCUSSION TITLE
[Dev Tools] Testing the shared bridge

### DISCUSSION THREAD

**Discussion by mesozoic-technology** [zkSync-Community-Hub/zksync-developers#781] [Dev Tooling] [answered]
### Environment

Testnet

### Select the Dev Tool you are using

Local Node

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

I am trying to figure out how to unit test the bridging of an erc20 token from l1 into a contract on l2. 

Ideally, I would have a hardhat test that can target deployment of my token contract on l1, and be able to trigger calls to that contract that will deposit these tokens into a contract on a zksync rollup. 

However it doesn't seem the hardhat plugin for zksync supports two nodes, it is just for the in-memory node which has no bridge between a given l1 and itself. 

How would you advise setting this up?

**Comment by alhenaai** [comment]
To set up a testing environment for bridging an ERC20 token from L1 to a contract on L2 using Hardhat, you'll need to simulate both L1 and L2 environments since the current Hardhat plugin for ZKsync primarily supports an in-memory node for L2 without a direct bridge to L1.

Here’s a suggested approach:

1. **Set Up Separate Hardhat Environments**: You'll need two separate Hardhat configurations: one for your L1 deployment and another for L2. This setup allows you to mimic the deployment on both layers independently.

2. **Use the Hardhat-zksync-node Plugin**: For L2, use the `hardhat-zksync-node` plugin to simulate the ZKsync environment. This plugin allows you to run tests against an in-memory ZKsync node. Configure this in your Hardhat setup for L2 testing.

3. **Mock the Bridging Logic**: Since the Hardhat environment does not natively support a bridge between L1 and L2, you might need to mock the bridging logic. This involves simulating the token transfer from L1 to L2 by manually adjusting token balances in your tests based on expected bridge behavior.

4. **Integration Testing**: For a more integrated approach, consider using scripts that manually handle the transition of state from L1 to L2. This can involve scripting the deployment on L1, simulating the bridging process, and then asserting the state on L2.

5. **Explore ZKsync's Testnet**: If feasible, testing directly on the ZKsync testnet might provide a more accurate representation of the bridging process, though this involves dealing with testnet deployment and potentially slower iteration cycles.

For more detailed guidance on setting up and using the Hardhat plugin for ZKsync, you can refer to the [Hardhat-zksync-node documentation](https://docs.zksync.io/build/tooling/hardhat/plugins/hardhat-zksync-node).

For further discussions and community support, you might find it helpful to visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by MexicanAce** [comment, marked as answer]
Hi @mesozoic-technology 👋 , if you need to have a test with both L1 and L2 in the loop, then I would recommend [using our Docker setup here](https://docs.zksync.io/build/test-and-debug/dockerized-l1-l2-nodes).

As you've pointed out, the in-memory node doesn't currently handle cross chain communication for testing, but the full docker setup should work for you 👍 (and the docs linked above will walk you through it). If you would also like more information about the differences between in-memory node and this docker test setup, [we have this document here](https://docs.zksync.io/build/test-and-debug).

**Reply by mesozoic-technology** [reply]
That's right the full docker setup has the backend with the l1 and the l2, but does it lend itself to only integration tests or both that and unit tests?

For a unit test, I'd need a clean slate in the docker setup at the beginning of each test. At first blush, it seems that the docker setup would result in a persistent node that would accrue the state of each transaction and to reset this, I'd have to tear it down and spin it up anew for each test. 

Would that be something I could encapsulate into a setup function in hardhat?

**Reply by mesozoic-technology** [reply]
I may be having issues with the docker setup, I'm getting a printout claiming the expected platform doesn't match the detected one.

https://github.com/matter-labs/local-setup/issues/30
